### ISSUE TITLE
win32 - file test operators don't work for //?/UNC/server/file filenames

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9025]
Migrated from [rt.perl.org#45331](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=45331) (status was 'open')

Searchable as RT45331$

**Comment by p5pRT** [comment]
### From adavies@ptc.com

#### Created by adavies@ptc.com



On win32 you can use UNC pathanmes of the form&#8203;:

\\\\?\\UNC\\server\\folder\\filename

Perl can C\<open> these files ok\, but C\<\-f>\, C\<\-e> etc\. 
incorrectly return false for these filenames\.
However\, they do work for directory names in this format\.

The code in question is in win32/win32\.c's win32\_stat\(\)&#8203;:

\# %\<

 \#if defined\(WIN64\) || defined\(USE\_LARGE\_FILES\)
&nbsp;    res = \_stati64\(path\, sbuf\);
 \#else
&nbsp;    res = stat\(path\, sbuf\);
 \#endif
&nbsp;    sbuf\->st\_nlink = nlink;

 if \(res \< 0\) \{
&nbsp;    /\* CRT is buggy on sharenames\, so make sure it really isn't\.
&nbsp;     \* XXX using GetFileAttributesEx\(\) will enable us to set
&nbsp;     \* sbuf\->st\_\*time \(but note that's not available on the
&nbsp;     \* Windows of 1995\) \*/
&nbsp;    DWORD r = GetFileAttributesA\(path\);
&nbsp;    if \(r \!= 0xffffffff && \(r & FILE\_ATTRIBUTE\_DIRECTORY\)\) \{
&nbsp;        /\* sbuf may still contain old garbage since stat\(\) failed \*/
&nbsp;        Zero\(sbuf\, 1\, Stat\_t\);
&nbsp;        sbuf\->st\_mode = S\_IFDIR | S\_IREAD;
&nbsp;        errno = 0;
&nbsp;        if \(\!\(r & FILE\_ATTRIBUTE\_READONLY\)\)
&nbsp;            sbuf\->st\_mode |= S\_IWRITE | S\_IEXEC;
&nbsp;        return 0;
&nbsp;    \}

\# >%


Note that GetFileAttributesA\(\) does succeed for "\\\\?\\UNC\\server\\file"
format filesnames\.\.\. but the result does not have the
"FILE\_ATTRIBUTE\_DIRECTORY" bit set\.

Is there any reason why this isn't coded as \(untested\)&#8203;:


&nbsp;    DWORD r = GetFileAttributesA\(path\);
&nbsp;    if \(r \!= 0xffffffff\) \{
&nbsp;        /\* sbuf may still contain old garbage since stat\(\) failed \*/
&nbsp;        Zero\(sbuf\, 1\, Stat\_t\);
&nbsp;        sbuf\->st\_mode = S\_IREAD;
&nbsp;        if \(r & FILE\_ATTRIBUTE\_DIRECTORY\)
&nbsp;        sbuf\->st\_mode |= S\_IFDIR;
&nbsp;        errno = 0;
&nbsp;        if \(\!\(r & FILE\_ATTRIBUTE\_READONLY\)\)
&nbsp;        sbuf\->st\_mode |= S\_IWRITE | S\_IEXEC;
&nbsp;        return 0;
&nbsp;    \}

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.8.7:

Configured by adavies at Thu Aug 11 14:02:10 2005.

Summary of my perl5 (revision 5 version 8 subversion 7) configuration:
  Platform:
    osname=MSWin32, osvers=5.1, archname=MSWin32-x86-multi-thread
    uname=''
    config_args='undef'
    hint=recommended, useposix=true, d_sigaction=undef
    usethreads=define use5005threads=undef useithreads=define
usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cl', ccflags ='-nologo -Gf -W3 -MD -DNDEBUG -O1 -DWIN32
-D_CONSOLE -DNO_STRICT -DHAVE_DES_FCRYPT  -DPERL_IMPLICIT_CONTEXT
-DPERL_IMPLICIT_SYS -DUSE_PERLIO -DPERL_MSVCRT_READFIX',
    optimize='-MD -DNDEBUG -O1',
    cppflags='-DWIN32'
    ccversion='12.00.8804', gccversion='', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=undef, longlongsize=8, d_longdbl=define, longdblsize=10
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='__int64',
lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='link', ldflags ='-nologo -nodefaultlib -release
-libpath:"c:\perl3\lib\CORE"  -machine:x86'
    libpth=\lib
    libs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib
comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib
netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib  version.lib
odbc32.lib odbccp32.lib msvcrt.lib
    perllibs=  oldnames.lib kernel32.lib user32.lib gdi32.lib
winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib
oleaut32.lib  netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib
version.lib odbc32.lib odbccp32.lib msvcrt.lib
    libc=msvcrt.lib, so=dll, useshrplib=yes, libperl=perl58.lib
    gnulibc_version='undef'
  Dynamic Linking:
    dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags='-dll -nologo -nodefaultlib -release
-libpath:"c:\perl3\lib\CORE"  -machine:x86'

Locally applied patches:
    


@INC for perl v5.8.7:
    C:/perl3/lib
    C:/perl3/site/lib
    .


Environment for perl v5.8.7:
    HOME=C:\alex
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
 
PATH=C:\WINNT\system32;C:\WINNT;C:\WINNT\System32\Wbem;C:\perl3\bin;D:\a
lex\bin;C:\cygwin\bin;C:\Program Files\Perforce;C:\Program
Files\Microsoft Visual Studio\VC98\Bin;C:\Program Files\Microsoft Visual
Studio\Common\MSDev98\Bin;C:\Program Files\Microsoft Visual
Studio\VC98\MFC\Lib;
    PERL_BADLANG (unset)
    SHELL (unset)

```
</details>

**Comment by p5pRT** [comment]
### From @jkeenan

On Mon Sep 10 09&#8203;:41&#8203;:21 2007\, adavies@&#8203;ptc\.com wrote&#8203;:
> This is a bug report for perl from adavies@&#8203;ptc\.com\,
> generated with the help of perlbug 1\.35 running under perl v5\.8\.7\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> On win32 you can use UNC pathanmes of the form&#8203;:
> 
> \\\\?\\UNC\\server\\folder\\filename
> 
> Perl can C\<open> these files ok\, but C\<\-f>\, C\<\-e> etc\. 
> incorrectly return false for these filenames\.
> However\, they do work for directory names in this format\.
> 
> The code in question is in win32/win32\.c's win32\_stat\(\)&#8203;:
> 
> \# %\<
> 
>  \#if defined\(WIN64\) || defined\(USE\_LARGE\_FILES\)
>     res = \_stati64\(path\, sbuf\);
>  \#else
>     res = stat\(path\, sbuf\);
>  \#endif
>     sbuf\->st\_nlink = nlink;
> 
>  if \(res \< 0\) \{
>     /\* CRT is buggy on sharenames\, so make sure it really isn't\.
>      \* XXX using GetFileAttributesEx\(\) will enable us to set
>      \* sbuf\->st\_\*time \(but note that's not available on the
>      \* Windows of 1995\) \*/
>     DWORD r = GetFileAttributesA\(path\);
>     if \(r \!= 0xffffffff && \(r & FILE\_ATTRIBUTE\_DIRECTORY\)\) \{
>         /\* sbuf may still contain old garbage since stat\(\) failed \*/
>         Zero\(sbuf\, 1\, Stat\_t\);
>         sbuf\->st\_mode = S\_IFDIR | S\_IREAD;
>         errno = 0;
>         if \(\!\(r & FILE\_ATTRIBUTE\_READONLY\)\)
>             sbuf\->st\_mode |= S\_IWRITE | S\_IEXEC;
>         return 0;
>     \}
> 
> \# >%
> 
> 
> Note that GetFileAttributesA\(\) does succeed for "\\\\?\\UNC\\server\\file"
> format filesnames\.\.\. but the result does not have the
> "FILE\_ATTRIBUTE\_DIRECTORY" bit set\.
> 
> Is there any reason why this isn't coded as \(untested\)&#8203;:
> 
> 
>     DWORD r = GetFileAttributesA\(path\);
>     if \(r \!= 0xffffffff\) \{
>         /\* sbuf may still contain old garbage since stat\(\) failed \*/
>         Zero\(sbuf\, 1\, Stat\_t\);
>         sbuf\->st\_mode = S\_IREAD;
>         if \(r & FILE\_ATTRIBUTE\_DIRECTORY\)
>         sbuf\->st\_mode |= S\_IFDIR;
>         errno = 0;
>         if \(\!\(r & FILE\_ATTRIBUTE\_READONLY\)\)
>         sbuf\->st\_mode |= S\_IWRITE | S\_IEXEC;
>         return 0;
>     \}
> 
> 
> 
> 

Can anyone on Win32 confirm whether these problems persist in a
supported version of Perl \(5\.14 or 5\.16\)?

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @tonycoz

On Mon\, Aug 13\, 2012 at 04&#8203;:10&#8203;:26PM \-0700\, James E Keenan via RT wrote&#8203;:
> On Mon Sep 10 09&#8203;:41&#8203;:21 2007\, adavies@&#8203;ptc\.com wrote&#8203;:
> > This is a bug report for perl from adavies@&#8203;ptc\.com\,
> > generated with the help of perlbug 1\.35 running under perl v5\.8\.7\.
> > 
> > 
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > \[Please enter your report here\]
> > 
> > On win32 you can use UNC pathanmes of the form&#8203;:
> > 
> > \\\\?\\UNC\\server\\folder\\filename
> > 
> > Perl can C\<open> these files ok\, but C\<\-f>\, C\<\-e> etc\. 
> > incorrectly return false for these filenames\.
> > However\, they do work for directory names in this format\.
> 
> Can anyone on Win32 confirm whether these problems persist in a
> supported version of Perl \(5\.14 or 5\.16\)?

This is still a problem in blead\.

C&#8203;:\\Users\\tony\\dev\\perl\\git\\perl>perl \-le "print qq\($\_&#8203;: \)\, join '\,'\, \-d $\_\, \-f $\_
\, \-e $\_ for @&#8203;ARGV" \\\\?\\UNC\\mars\\tony\\test\.png \\\\?\\UNC\\mars\\tony\\library \\\\mars\\t
ony\\test\.png \\\\mars\\tony\\library
\\\\?\\UNC\\mars\\tony\\test\.png&#8203;: \,\,
\\\\?\\UNC\\mars\\tony\\library&#8203;: 1\,\,1
\\\\mars\\tony\\test\.png&#8203;: \,1\,1
\\\\mars\\tony\\library&#8203;: 1\,\,1

Tony

**Comment by tonycoz** [comment]
Fixed in blead, by e935ef333b3eab54a766de93fad1369f76ddea49:

```
C:\Users\Tony\dev\perl\git\perl\win32>..\perl -le "print qq($_: ), join ',', -d $_, -f $_, -e $_ for @ARGV" \\?\UNC\mars\tony\test.png \\?\UNC\mars\tony\library \\mars\tony\test.png \\mars\tony\library
\\?\UNC\mars\tony\test.png: ,1,1
\\?\UNC\mars\tony\library: 1,,1
\\mars\tony\test.png: ,1,1
\\mars\tony\library: 1,,1
```

I suspect a build with a recent MSVC would also fix it, since the UCRT uses CreateFile() to open file for stat() rather than FindFirstFile() as older MSVC and the msvcrt do.
